src/argparse.zig

```diff
@@ -0,0 +1,424 @@
+					const arg = nullableArg orelse return missingArgument(Field, allocator, name);
+					return .{.success = std.fmt.parseFloat(std.meta.Float(floatInfo.bits), arg) catch {
+						return .initWithFailure(allocator, utils.format(allocator, "Expected a number for {s}, found \"{s}\"", .{name, arg}));
+					}};
+				},
+				inline .int => |intInfo| {
+					const arg = nullableArg orelse return missingArgument(Field, allocator, name);
+					return .{.success = std.fmt.parseInt(std.meta.Int(intInfo.signedness, intInfo.bits), arg, 0) catch {
+						return .initWithFailure(allocator, utils.format(allocator, "Expected an integer for {s}, found \"{s}\"", .{name, arg}));
+					}};
+				},
+				inline else => |other| @compileError("Unsupported type " ++ @tagName(other)),
+			}
+		}
+
+		fn missingArgument(comptime Field: type, allocator: NeverFailingAllocator, name: []const u8) ParseResult(Field) {
+			return .initWithFailure(allocator, utils.format(allocator, "Missing argument at position {s}", .{name}));
+		}
+
+		fn autocompleteArgument(comptime Field: type, allocator: NeverFailingAllocator, _arg: ?[]const u8) AutocompleteResult {
+			const arg = _arg orelse return .{};
+			switch(@typeInfo(Field)) {
+				inline .@"struct" => {
+					if(!@hasDecl(Field, "autocomplete")) @compileError("Struct must have an autocomplete function");
+					return try @field(Field, "autocomplete")(allocator, arg);
+				},
+				inline .@"enum" => {
+					var result: AutocompleteResult = .{};
+					inline for(std.meta.fieldNames(Field)) |fieldName| {
+						if(!std.mem.startsWith(u8, fieldName, arg)) continue;
+						result.suggestions.append(allocator, allocator.dupe(u8, fieldName));
+					}
+					return result;
+				},
+				inline else => return .{},
+			}
+		}
+
+		fn parseUnion(comptime u: std.builtin.Type.Union, allocator: NeverFailingAllocator, args: []const u8) ParseResult(T) {
+			var result: ParseResult(T) = .initWithFailure(allocator, allocator.dupe(u8, "Couldn't match argument list."));
+
+			inline for(u.fields) |field| {
+				var fieldResult = Parser(field.type, options).resolve(false, allocator, args);
+				defer fieldResult.deinit(allocator);
+
+				if(fieldResult == .success) {
+					result.deinit(allocator);
+					return .{.success = @unionInit(T, field.name, fieldResult.success)};
+				}
+
+				var messageSegments: ListUnmanaged([]const u8) = .{};
+				defer messageSegments.deinit(allocator);
+				var maxSize: usize = 0;
+
+				messageSegments.append(allocator, options.commandName);
+				maxSize += options.commandName.len;
+
+				for(std.meta.fieldNames(field.type)) |name| {
+					messageSegments.append(allocator, " ");
+					messageSegments.append(allocator, name);
+					maxSize += name.len;
+				}
+
+				const message = std.mem.concat(allocator.allocator, u8, messageSegments.items) catch unreachable;
+				result.failure.messages.append(allocator, message);
+				result.failure.takeMessages(allocator, &fieldResult.failure);
```

Comment:
You have one member function here, and one function which accesses the messages field directly.
So why not just do the same thing here as well and remove the then unused member function?
```suggestion
				result.failure.messages.appendSlice(allocator, fieldResult.failure.messages.items);
```